Adds recording mode spec. Related to #93 (#155)
microsoftgraph/msgraph-developer-proxy: specs/recording-requests.md

> # Recording requests

Reviewer: ```suggestion
# Record request logging
```

> Recording requests allows Developer Proxy to capture requests and related messages that were intercepted by the proxy while in record mode. The captured requests are passed for further processing to the Developer Proxy plugins. This feature is foundational for scenarios such as permission analysis, building tailored SDKs, audit reports, etc.

Reviewer: Recording is such an overloaded term, I got this idea confused with the idea of recording requests and responses to generate a mock responses file.

Perhaps we should consider being more explicit in the title and opening that we're recording the logging data?

Author: This would be totally possible with a dedicated recording plugin! 😀 

Would capturing be clearer?

Reviewer: Perhaps we call this Record request logging, something like:
```suggestion
Record request logging allows Developer Proxy to capture request logging activities that occurred while the proxy was in record mode. The captured request logging items are passed for further processing to the Developer Proxy plugins. This feature is foundational for scenarios such as permission analysis, building tailored SDKs, audit reports, etc.
```

## History

| Version | Date | Comments | Author |
| ------- | ---- | -------- | ------ |
| 1.0 | 2023-02-01 | Initial specifications | @waldekmastykarz |

## Implementation

> Developer Proxy can be put in recording mode either by using the `--record` command line option or by pressing `r` after starting the proxy. Recording can be stopped by pressing `s` while running the proxy or by closing the proxy using `CTRL+C`. When the recording is stopped while proxy is running, users can start a new recording session by pressing `r` again.

Reviewer: ```suggestion
Developer Proxy can be put in recording mode either by using the `--record` command line option or by pressing `r` after starting the proxy. Recording can be stopped by pressing `s` while running the proxy or by closing the proxy using `CTRL+C`. While recording is stopped and the proxy is running, users can start a new recording session by pressing `r` again.
```

> While recording, Developer Proxy will show and indicator in the terminal window. The indicator will be displayed in the top right corner of the window and will be hidden if the proxy is not recording.

Reviewer: ```suggestion
While recording, Developer Proxy will show an indicator in the terminal window. The indicator will be displayed in the top right corner of the window and will be hidden if the proxy is not recording.
```

> Proxy is recording requests by subscribing to a `LogRequest` event which is raised each time proxy logs a request-related log message. The event contains information about the logged message.

Reviewer: Is this event to be raised on each call to `ILoger.LogRequest`?
Does it happen before the logging happens or after?

Author: Yes, it's this event. I suggest we raise the event after the message was logged. It aligns more with the "Logged" aspect (past tense).

Reviewer: Perhaps a slight tweak to the event name `RequestLogged` gets this to the place we want it.

IMO, for the moment we leave it and address when/if we focus on custom loggers.

> After the recording is stopped, Developer Proxy will raise the `RecordingStopped` event. Plugins can subscribe to this event to process the captured requests.

Reviewer: If I'm understanding this idea correctly we could have a new kind of plugin, say a `SessionSummaryPlugin` that on `RecordingStopped` would receive all of the collected logged request data and messages. This plugin could then generate some output based on the data it receives.

Author: Correct, so this gives us a lot of flexibility when it comes to processing a set of requests, like summaries, analysis or passing on to other tools like Kiota.
